client/telemetry/src/worker.rs

```diff
@@ -115,19 +122,36 @@ impl TelemetryWorker {
 			CONNECT_TIMEOUT
 		).boxed();
 
+		let mut force_connected = false;
+		let nodes = endpoints.into_iter().map(|(addr, verbosity)| {
+			let node = if let Some(node_pool) = node_pool {
+				let (node, is_new) = node_pool.get_or_create(transport.clone(), addr);
+				if !is_new {
+					force_connected = true;
+				}
+				node
+			} else {
+				Arc::new(Mutex::new(node::Node::new(transport.clone(), addr)))
+			};
+			(node, verbosity)
+		}).collect();
```

Comment:
I wonder if we can do this work a bit higher up the call-stack to avoid passing around locks? I don't have a concrete suggestion/criticism here, just a feeling that this feels more complex than it needs to be?

Reply:
Thanks! Fixed in a8c6e8cd1

I checked again why I needed a `Arc<Mutex<...>>` there and I realized I could remove it if I create the `Telemetries` first instead of the `TelemetryLayer` first (which requires Send and Sync).

Here was my reasoning:

The current way it works is (from bottom to top):
 - `Node`: an actual connection to a telemetry server
 - `Worker`: has multiple `Node`, multiplexes messages by sending them to all the nodes
 - `Telemetry`: wraps a worker with a channel receiver and offers a Stream that pass the messages from the receiver to the worker

What I am adding:
 - `NodePool`: a pool of `Node` which I can get-or-create
 - `Senders`: a collection of channel senders which is used by `TelemetryLayer` to route the log to the correct channel based on the span's id
 - `Telemetries`: wraps `Senders` and `NodePool`. The nodes creates from `Telemetries` re-uses `Node` using the `NodePool`. The `Senders` is a shared resource with the `TelemetryLayer` that help routing the logs to the correct channel based on the span's id
- `TelemetryLayer`: that's the tracing layer that handles the logs for telemetries. It wraps `Telemetries` and mostly only access `senders`. It also allow retrieving its `Telemetries` object which is used in `sc_service::Configuration` to create the `Telemetry` instances

We could reverse the way `Telemetries` and `TelemetryLayer` are created. Currently the layer is created and the telemetries are cloned/extracted from it. We could create the telemetries instead and have a method that would generate the layer. Or we could combine the layer and telemetries into a single entity but I'm afraid it will makes things confusing.

Reply:
Thank you for the detailed explanation. `a8c6e8c` looks much better. :)